src/Signing/SignatureVerifier.php

```diff
@@ -47,12 +55,20 @@ public function __construct(KeyLoader $keyLoader)
      */
     public function hasValidSignature(AuthnRequest $request, ServiceProvider $serviceProvider)
     {
+        $this->logger->debug(sprintf('Extracting public keys for ServiceProvider "%s"', $serviceProvider->getEntityId()));
+
         $keys = $this->keyLoader->extractPublicKeys($serviceProvider);
 
+        $this->logger->debug(sprintf('Found "%d" keys, filtering the keys to get X509 keys', $keys->count()));
         $x509Keys = $keys->filter(function (SAML2_Certificate_Key $key) {
             return $key instanceof SAML2_Certificate_X509;
         });
 
+        $this->logger->debug(sprintf(
+            'Found "%d" X509 keys, attempting to use each for signature verification',
+            $x509Keys->count()
+        ));
+
         foreach ($x509Keys as $key) {
             if ($this->isSignedWith($request, $key)) {
                 return true;
```

Comment:
That is actually done in the `isSignedWith` method. If all keys fail, false is returned and that immediately triggers an exception. Not really sure if it adds anything to duplicate the `Signature VERIFIED` below here again, will add a statement that the signature could not be verified with any key.